dso/model.py

```diff
@@ -431,6 +438,29 @@ def key(self) -> str:
         return _as_key(self.provider_id, self.ruleset_id, self.rule_id)
 
 
+@dataclasses.dataclass(frozen=True)
```

Comment:
nit: I suppose you added `frozen=True` for consistency w/ the other classes in this module. however, as it has a mutable attr (resources), it cannot truly be frozen (which is why I suggest to drop `frozen=True` param